File: temp_brace_check.py
from pathlib import Path
text = Path('app/Controllers/Admin.php').read_text()
line = 1
stack = []
for idx, ch in enumerate(text, 1):
    if ch == '\n':
        line += 1
    elif ch == '{':
        stack.append((idx, line))
    elif ch == '}':
        if stack:
            stack.pop()
        else:
            print('extra_close', idx, line)
if stack:
    print('unclosed_open_count', len(stack))
    for pos, l in stack[-10:]:
        print('unclosed_open_at', pos, l)
else:
    print('balanced')
print('total opens', text.count('{'), 'total closes', text.count('}'))

File: temp_brace_check.php
<?php
$code = file_get_contents('app/Controllers/Admin.php');
$tokens = token_get_all($code);
$stack = [];
$line = 1;
foreach ($tokens as $token) {
    if (is_array($token)) {
        $text = $token[1];
        $line += substr_count($text, "\n");
        if ($token[0] === T_COMMENT || $token[0] === T_DOC_COMMENT || $token[0] === T_CONSTANT_ENCAPSED_STRING || $token[0] === T_ENCAPSED_AND_WHITESPACE) {
            continue;
        }
        $text = $token[1];
    } else {
        $text = $token;
    }
    for ($i = 0; $i < strlen($text); $i++) {
        $ch = $text[$i];
        if ($ch === '{') {
            $stack[] = ['line' => $line, 'char' => $ch];
        } elseif ($ch === '}') {
            if ($stack) {
                array_pop($stack);
            } else {
                echo "extra_close at line $line\n";
            }
        }
    }
}
if ($stack) {
    echo "unclosed_open_count " . count($stack) . "\n";
    foreach ($stack as $item) {
        echo "unclosed_open_at line " . $item['line'] . "\n";
    }
} else {
    echo "balanced\n";
}
echo "total opens " . count(array_filter($stack, fn($x)=>true)) . "\n";
